### PR TITLE
Bug 1876217 - Add full keyword support for AMP items

### PR DESCRIPTION
**Pull request by bendk**
Parse the `full_keyword` field for downloaded AMP / wikipedia suggestions and store it in the DB.

Use the field for AMP suggestions rather than the current hueristic-based code.

Don't use this field for Wikimedia suggestions.  This is because the full keyword field is just the full article title, which is already included in the suggestion.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [x] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/suggest/src/store.rs [hunk 1]
```diff
@@ -998,6 +998,211 @@ mod tests {
...
+                        impression_url: "https://example.com/impression_url",
+                        click_url: "https://example.com/click_url",
+                        raw_click_url: "https://example.com/click_url",
+                        score: 0.3,
+                    },
+                ]
+            "#]]
+            .assert_debug_eq(&dao.fetch_suggestions(&SuggestionQuery {
+                keyword: "los pollos h".into(),
+                providers: vec![SuggestionProvider::Amp],
```
Comment (bendk) [inline components/suggest/src/store.rs:1179]: Should I also be handling `SuggestionProvider::AmpMobile`?  I'm not quite sure what the relationship between the two AMP providers is.

File: components/suggest/src/db.rs [hunk 2]
```diff
@@ -853,22 +859,56 @@ impl<'a> SuggestDao<'a> {
...
+                                 VALUES(
+                                    :suggestion_id,
+                                    :keyword
+                                 )",
+                                named_params! {
+                                    ":keyword": full_keyword,
+                                    ":suggestion_id": suggestion_id,
+                                },
+                            )?;
+                            let full_keyword_id = self.conn.last_insert_rowid();
```
Comment (ncloudioj) [inline components/suggest/src/db.rs:886]: We often use `RETURNING id` in the `Insert` SQL to get the last insert rowid, can we do that here as well?

**Review by ncloudioj — APPROVED** [review]
Looks great, only a few minor comments for your consideration.

File: components/suggest/src/store.rs (hunk 1 above)
Comment (bendk) [reply components/suggest/src/store.rs:1210]: That makes sense, I added a test and refactored some code for this.  Can you review that the changes make sense?

File: components/suggest/src/rs.rs [hunk 3]
```diff
@@ -281,6 +283,41 @@ impl DownloadedAmpWikipediaSuggestion {
...
+
+            AmpKeyword {
+                rank: i,
+                keyword,
+                full_keyword: common
+                    .full_keywords
+                    .get(full_keyword_index)
+                    .map(|(keyword, _)| keyword.as_str()),
+            }
+        })
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/suggest/src/rs.rs]: Really slick, love it.

File: components/suggest/src/db.rs (hunk 2 above)
Comment (bendk) [reply components/suggest/src/db.rs]: TIL.  I updated the code to use that.

File: components/suggest/src/db.rs [hunk 4]
```diff
@@ -1339,6 +1366,53 @@ impl<'a> SuggestDao<'a> {
...
+impl<'a> FullKeywordInserter<'a> {
+    fn new(conn: &'a Connection, suggestion_id: i64) -> Self {
+        Self {
+            conn,
+            suggestion_id,
+            last_inserted: None,
+        }
+    }
+
+    fn get_id(&mut self, full_keyword: &'a str) -> rusqlite::Result<i64> {
```
Comment (ncloudioj) [inline components/suggest/src/db.rs:1389]: Suggestion: how about renaming this to `maybe_insert()`? `get_id()` kind of hides the insert part.

**Review by ncloudioj — APPROVED** [review]
Still r+!

**Pull request merged**